File: mobile/lib/models/animal.dart
/// Animal model for the mobile app.
class AnimalModel {
  final int id;
  final String animalUid;
  final String? name;
  final String species;
  final String? breed;
  final String gender;
  final double? weight;
  final String? growthStage;
  final String status;
  final String ownerName;
  final String? notes;
  final String createdAt;
  bool isSynced;

  AnimalModel({
    required this.id,
    required this.animalUid,
    this.name,
    required this.species,
    this.breed,
    required this.gender,
    this.weight,
    this.growthStage,
    required this.status,
    required this.ownerName,
    this.notes,
    required this.createdAt,
    this.isSynced = true,
  });

  factory AnimalModel.fromJson(Map<String, dynamic> json) {
    return AnimalModel(
      id: json['id'],
      animalUid: json['animal_uid'],
      name: json['name'],
      species: json['species'],
      breed: json['breed'],
      gender: json['gender'],
      weight: json['weight']?.toDouble(),
      growthStage: json['growth_stage'],
      status: json['status'],
      ownerName: json['owner_name'] ?? '',
      notes: json['notes'],
      createdAt: json['created_at'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'animal_uid': animalUid,
      'name': name,
      'species': species,
      'breed': breed,
      'gender': gender,
      'weight': weight,
      'growth_stage': growthStage,
      'status': status,
      'owner_name': ownerName,
      'notes': notes,
      'created_at': createdAt,
      'is_synced': isSynced,
    };
  }
}

File: backend/app/models/animal.py
"""Animal model — core entity for livestock tracking."""
from sqlalchemy import Column, Integer, String, Float, Date, DateTime, ForeignKey, Text
from sqlalchemy.orm import relationship
from datetime import datetime, timezone
from app.database import Base


class Animal(Base):
    __tablename__ = "animals"

    id = Column(Integer, primary_key=True, index=True)
    animal_uid = Column(String(50), unique=True, nullable=False, index=True)
    name = Column(String(100), nullable=True)
    species = Column(String(50), nullable=False)  # cattle, goat, sheep, pig, poultry
    breed = Column(String(100), nullable=True)
    date_of_birth = Column(Date, nullable=True)
    gender = Column(String(10), nullable=False)  # male, female
    weight = Column(Float, nullable=True)
    growth_stage = Column(String(50), nullable=True)  # weaner, grower, finisher, breeder
    status = Column(String(30), default="active")  # active, sold, deceased, transferred
    owner_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    registered_by = Column(Integer, ForeignKey("users.id"), nullable=False)
    qr_code_path = Column(String(500), nullable=True)
    notes = Column(Text, nullable=True)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime, default=lambda: datetime.now(timezone.utc),
                        onupdate=lambda: datetime.now(timezone.utc))

    # Relationships
    owner = relationship("User", back_populates="owned_animals",
                         foreign_keys=[owner_id])
    registered_by_user = relationship("User", back_populates="registered_animals",
                                       foreign_keys=[registered_by])
    health_records = relationship("HealthRecord", back_populates="animal",
                                   cascade="all, delete-orphan")
    treatments = relationship("Treatment", back_populates="animal",
                               cascade="all, delete-orphan")
    vaccinations = relationship("Vaccination", back_populates="animal",
                                 cascade="all, delete-orphan")
    movements = relationship("Movement", back_populates="animal",
                              cascade="all, delete-orphan")
    qr_code = relationship("QRCode", back_populates="animal", uselist=False,
                            cascade="all, delete-orphan")

    def __repr__(self):
        return f"<Animal(id={self.id}, uid='{self.animal_uid}', species='{self.species}')>"
